Move existing Admin pages/components to align with new Sidebar setup

misicnenad opened the pull request:
Ticket(s):
- https://github.com/AngelProtocolFinance/angelprotocol-web-app/issues/1720

## Explanation of the solution
Extending from https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/1716

## Instructions on making this work

- run `yarn` or `yarn install` to install npm dependencies
- run `yarn run test --watchAll` to verify all tests still pass
- (optional) run `yarn run build` to verify the build passes
- run `yarn start` to start the webapp
- connect Charity 1 wallet
- go to admin page(s) http://localhost:4200/admin/1
- verify all admin pages now use the new Sidebar component + layout on both desktop and mobile

## UI changes for review

_**NOTE: there were issues loading endowments due to changes to the `type EndowmentProfile`, so the errors you may encounter are not due to changes in this PR.**_

### Desktop
#### Light
![image](https://user-images.githubusercontent.com/19427053/220837127-7c68c984-3ba9-4dfa-a10e-828c3f46c915.png)
#### Dark
![image](https://user-images.githubusercontent.com/19427053/220837149-071f9dd3-467a-4215-a498-dd0ce5c8a1ba.png)

### Mobile
#### Light
![image](https://user-images.githubusercontent.com/19427053/220837309-1dc04fb7-1ef7-4338-a5a8-a8015135172b.png)
![image](https://user-images.githubusercontent.com/19427053/220837327-48c7f06d-f106-48bc-9bef-7b683cc8e46c.png)
#### Dark
![image](https://user-images.githubusercontent.com/19427053/220837258-bdcb5862-6c5a-4c63-bd21-fa0097067234.png)
![image](https://user-images.githubusercontent.com/19427053/220837290-2d71a16d-2e7e-4c50-af95-8b16f1bc35ca.png)

Button title on mobile screen for subpaths (e.g. `Settings` appears as the selected link even when the current path is a subpath the selected link, in this case the user is on `admin/1/settings/edit/liquid`):
![image](https://user-images.githubusercontent.com/19427053/220839475-a7c3ecdc-418d-4665-b410-89c8f6398a3b.png)

misicnenad commented on src/pages/Admin/Charity/index.tsx:
```diff
@@ -15,20 +19,19 @@ import Withdraws from "./Withdraws";
 export default function Charity() {
+  const { id } = useAdminResources();
+
+  const linkGroups = useMemo(() => createLinkGroups(id), [id]);
```
Avoids recreating the list on every render.

misicnenad commented on src/pages/Admin/constants.ts:
```diff
@@ -40,3 +42,173 @@ export const proposalShape: SchemaShape<ProposalBase> = {
+};
+
+// THESE WILL BE THE APPROPRIATE ICONS TO USE FOR FUTURE PAGES
+// TODO: REMOVE ONCE ALL ARE MOVED INTO `LINKS` ABOVE.
```
Left this here as it will be helpful once we get to creating/migrating the pages with icons/URLs defined within.

misicnenad commented on src/pages/Admin/Sidebar/useMobileSidebar.tsx:
```diff
@@ -0,0 +1,50 @@
+  to: "",
+};
+
+export default function useMobileSidebar(linkGroups: LinkGroup[]) {
```
Component present, see [line 38](https://github.com/AngelProtocolFinance/angelprotocol-web-app/blob/ee694b640e49e149b79186ca1dc84c1beebc67f9/src/pages/Admin/Sidebar/useMobileSidebar.tsx#L38)

ap-justin replied:
component should not be present in hook, should be declared in it's own `.tsx` file

misicnenad replied:
Will update.

misicnenad commented on src/pages/Admin/Charity/index.tsx:
```diff
@@ -51,3 +54,30 @@ export default function Charity() {
   );
 }
+
+function createLinkGroups(endowId: number): LinkGroup[] {
```
The function is necessary to append `endowId` to end of `adminRoutes.widget_config` path.

misicnenad replied:
It is also run only once (due to `useMemo`), so no performance overhead.

misicnenad commented on src/pages/Admin/Charity/index.tsx:
```diff
@@ -51,3 +54,30 @@ export default function Charity() {
+        LINKS[adminRoutes.proposals],
+        {
+          ...LINKS[adminRoutes.widget_config],
+          to: `${LINKS[adminRoutes.widget_config].to}/${endowId}`,
```
`/{endowId}` *is* necessary for the widget configurer to know what is the default endowment to create the widget for, see PR https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/1746 for more details.

ap-justin replied:
if `<WidgetConfigurer/>` needs to know about default `endowmentId`, could just get it by itself 
```ts
//WidgetConfigurer
const { id } = useAdminResources()
```

misicnenad replied:
Yes, but then there would be no way to tell the widget that the endowment to donate to will be selected using its (widget's) combobox:
- if `/{endowId}` is present -> just display the endowment name
- if no ID param is present -> display the endowment combobox

misicnenad replied:
Anyway, that's how the widget configurer currently works. Any changes to its function are out of scope for this PR.

ap-justin replied:
@misicnenad, you are using this `<WidgetConfigurer/>` in the context of `/admin/:id`
just look at how you use your `createLinkGroups` - literally just passing `id` of current admin page 

```ts
 const { id } = useAdminResources();
 const linkGroups = useMemo(() => createLinkGroups(id), [id]); 
  
```

> Yes, but then there would be no way to tell the widget that the endowment to donate to will be selected using its (widget's) combobox:

there's even no need to for `configurer/{endowId}` 
If I am currently in `admin/1` is there a way for me to navigate to `admin/{1}/widget-configurer/{2}` without explicitly typing it in a browser?

> if no ID param is present -> display the endowment combobox

there's always `ID` param, in `admin/:id`, and `<WidgetConfigurer/>` should display the endowment's name by default. If user intends to configure the widget to donate to other endowment - could just select other in dropdown.


> Anyway, that's how the widget configurer currently works. Any changes to its function are out of scope for this PR.

imo, we should adjust a component's `api` to suite the page/use case it is used into - not adjust page implementation just to suit already merge component `api`

misicnenad replied:
> If I am currently in admin/1 is there a way for me to navigate to admin/{1}/widget-configurer/{2} without explicitly typing it in a browser?

That's a valid point @SovereignAndrey 

> there's always ID param, in admin/:id, and <WidgetConfigurer/> should display the endowment's name by default. If user intends to configure the widget to donate to other endowment - could just select other in dropdown

This goes against the original requirements https://app.clickup.com/t/865bp9483?comment=90040003682783&threadedComment=90080010341495

> there's always ID param, in admin/:id, and <WidgetConfigurer/> should display the endowment's name by default. If user intends to configure the widget to donate to other endowment - could just select other in dropdown
> ...
> imo, we should adjust a component's api to suite the page/use case it is used into - not adjust page implementation just to suit already merge component api

If we're talking about an internal implementation, then I agree with this statement.
But when we're talking about something external. like UI/UX (whether an endowment selector or plain text of their name appears), I'd be more careful with these changes and generally leave these types of decisions to the product owners - in these cases, the changes are best left for a separate PR.

misicnenad replied:
@SovereignAndrey @ap-justin **this PR is just about updating the existing admin pages _as they are_ to use the new Sidebar. Whether the specific pages' logic should be updated, although a valid concern, should be treated as a completely separate issue.**

ap-justin replied:
@misicnenad 

> This goes against the original requirements ([link to comment](https://app.clickup.com/t/865bp9483?comment=90040003682783&threadedComment=90080010341495)) and is therefore out of scope of this PR.

I believe this is for a widget being used as a standalone page `/widget-configurer/:id`

but in the context of an `admin/endowment` - an admin member should only concern itself on configuring their endowment's widget as you mentioned `A charity shouldn't need to have to navigate to Widget configurer with a charity that's not them set as default`

 that is 

```tsx
//widget-configurer used in the context of an admin
const { id } = useAdminResources()
<WidgetConfigurer defaultid={id} /> 
```

if widget configurer is to be used as a standalone page - like `user/create-widget` then param` endowId` makes sense and could be implemented

```tsx
//create widget page
const {endowIdFromParam} = useParams()
<WidgetConfigurer defaultid={endowIdFromParam} /> 
```

I am questioning the validity of `endowId` param for the use in the admin pages

so to avoid creating a `createLinkGroups` helper that just reassigns/reshuffles `constant LINKS` values 
```
        LINKS[adminRoutes.proposals],
        {
          ...LINKS[adminRoutes.widget_config],
          to: `${LINKS[adminRoutes.widget_config].to}/${endowId}`,
```
and recommends to just use `LINKS` directly to unneccesary layers of abstraction

misicnenad replied:
@ap-justin this is just how the current implementation of widget configurer is used. Updating the widget configurer **is out of scope of this PR. This PR only concerns itself with updating the existing pages as they are to use the new Sidebar**. Lets create a new ticket to update the widget configurer and move this discussion there.

ap-justin replied:
@misicnenad, since widget-configurer forces to make unnecessary layers of reassignment in the implementation of this layout

```ts
//createLinkGroup
  return [
    {
      links: [
        LINKS[adminRoutes.index],
        LINKS[adminRoutes.withdraws],
        LINKS[adminRoutes.contributions],
      ],
    },
    {
      title: "Invest",
      links: [LINKS[adminRoutes.invest], LINKS[adminRoutes.settings]],
    },
    { title: "Profile", links: [LINKS[adminRoutes.edit_profile]] },
    {
      title: "Manage",
      links: [
        LINKS[adminRoutes.proposals],
        {
          ...LINKS[adminRoutes.widget_config],
          to: `${LINKS[adminRoutes.widget_config].to}/${endowId}`,
```
I suggest we don't include it in this PR and layout components be implemented to just use plain `const LINKS`

misicnenad replied:
And I suggest we do that as part of a separate PR. @ap-justin please, this is not relevant to the task at hand and the implementation is not that complex either to execute or to understand. This is not that big of a deal that it has to hold back the whole PR.

ap-justin replied:
> If you want to discuss how to optimize widget configurer

I don't know what you mean by `optimize widget-configurer` I just want these code not merged in this `<Layout/>` implementation`
```
//createLinkGroup
  return [
    {
      links: [
        LINKS[adminRoutes.index],
        LINKS[adminRoutes.withdraws],
        LINKS[adminRoutes.contributions],
      ],
    },
    {
      title: "Invest",
      links: [LINKS[adminRoutes.invest], LINKS[adminRoutes.settings]],
    },
    { title: "Profile", links: [LINKS[adminRoutes.edit_profile]] },
    {
      title: "Manage",
      links: [
        LINKS[adminRoutes.proposals],
        {
          ...LINKS[adminRoutes.widget_config],
          to: `${LINKS[adminRoutes.widget_config].to}/${endowId}`,
```

> optimize widget configurer, that'd be great, but lets do that as part of a separate and relevant task/PR, not inside a task to add sidebar to existing pages.

that's why I'am asking not to include `<WidgetConfigurer/>` since its implementation is just for standalone page [see comment](https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/1804#discussion_r1112635054)

ap-justin replied:
@misicnenad - see simple solution #1846

misicnenad replied:
> that's why I'am asking not to include `<WidgetConfigurer/>`

@ap-justin this is not my decision to make, you should talk to @SovereignAndrey

misicnenad replied:
Will remove link to widget configurer for now, as it was commented out on `RC` branch anyway.

misicnenad commented on src/pages/Admin/Sidebar/Header.tsx:
```diff
@@ -2,9 +2,11 @@ import { useProfileQuery } from "services/aws/aws";
-  const { data: profile, isLoading, isError } = useProfileQuery(endowId);
+export default function Header() {
+  const { id } = useAdminResources();
+  const { data: profile, isLoading, isError } = useProfileQuery(id);
```
@SovereignAndrey do we hide sidebar header when the connected wallet belongs to either AP or Review team?

SovereignAndrey replied:
@misicnenad @ap-justin Sidebar header would NOT be hidden for AP Team nor for the Review Team. In reality, we'll probably want to have a set header that would act as an extra cue for those users as to which Admin area they are in and avoid accidental proposals to the wrong contracts. An example cue might be a solid header background color that is different for each ap admin type.

misicnenad replied:
@SovereignAndrey what would we display in Header for AP and Review teams?
- Logo: can be displayed -> could add some additional text to it as extra cue, like "Endowment", "AP Team", "Review".
- Endowment name -> makes little sense for them, do we display something else here?
- Endowment address -> what would we display here? 

All in all, we could have a header with different background colors and maybe difference logos depending on what type of wallet is connected.

ap-justin replied:
@misicnenad 

> @misicnenad @ap-justin Sidebar header would NOT be hidden for AP Team nor for the Review Team.

in the mininum, kindly apply these changes 
> so useProfileQuery should be skipped if adminResource.type !== "charity"

since `AP` and `REVIEW` ids doesn't pertain to any endowment-id

misicnenad replied:
@ap-justin will update, just waiting on Andrey's comments on the header subcomponents described above.

SovereignAndrey replied:
> @SovereignAndrey what would we display in Header for AP and Review teams?
> logo 
Use a simple placeholder like angelwings image?

> - Endowment name -> makes little sense for them, do we display something else here?
Nothing here, unless here's where we would put the above mentioned extra text: `"Angel Giving Team Admin", "Charity Applications Review".`

> - Endowment address -> what would we display here? 
Nothing needs to go here for now. 

> 
> All in all, we could have a header with different background colors and maybe difference logos depending on what type of wallet is connected. 
👍

misicnenad replied:
@SovereignAndrey different color examples (subject to change, these were just the first colors that popped to mind, not much thought went into them):
- Charities - as-is:
![image](https://user-images.githubusercontent.com/19427053/220546318-512e2db5-027f-49aa-8715-ea117da27866.png)
![image](https://user-images.githubusercontent.com/19427053/220546288-f5aa11b1-364f-4dd6-a381-2e00fff5b5c4.png)
- AP team:
![image](https://user-images.githubusercontent.com/19427053/220545931-52ceded6-04c0-41dd-8119-5c6add579642.png)
![image](https://user-images.githubusercontent.com/19427053/220545958-5da20cbe-617a-489f-9d0d-41d95cfb95a5.png)
- Review team:
![image](https://user-images.githubusercontent.com/19427053/220545826-3dbff471-2884-4879-8347-5b9cac341279.png)
![image](https://user-images.githubusercontent.com/19427053/220545850-40449e64-a641-4832-ae70-bbf508249b3d.png)

SovereignAndrey replied:
> * AP team:
>   ![image](https://user-images.githubusercontent.com/19427053/220545931-52ceded6-04c0-41dd-8119-5c6add579642.png)

The yellow is a bit jarring. Let's make both admins' colors consistent, regardless of light/dark mode. I would use the below dark mode colors for both light & dark receptively. 

> * AP team:
>   ![image](https://user-images.githubusercontent.com/19427053/220545958-5da20cbe-617a-489f-9d0d-41d95cfb95a5.png)
> * Review team:
>   ![image](https://user-images.githubusercontent.com/19427053/220545850-40449e64-a641-4832-ae70-bbf508249b3d.png)

misicnenad replied:
Updated so that all headers have same background color on light mode, but dark colors are as in last @SovereignAndrey's comment .

ap-justin commented on src/pages/Admin/Charity/Dashboard/index.tsx:
```diff
@@ -28,7 +27,6 @@ export default function Dashboard() {
       >
         {(proposals) => <Table proposals={proposals} />}
       </QueryLoader>
-      <Donations classes="mt-10" />
```
contributions should also exists in dashboard
<img width="809" alt="image" src="https://user-images.githubusercontent.com/89639563/220275042-281602c2-236e-4d28-b297-54396b67528b.png">

ap-justin commented on src/pages/Admin/Review/index.tsx:
```diff
@@ -1,22 +1,22 @@
 import Templates from "./Templates";
 
+const LINK_GROUPS: LinkGroup[] = [
+  { links: [LINKS[adminRoutes.index]] },
```
`admin/revew` index route `/` should be called `"Applications"` not `"Dasboard"`

ap-justin commented on src/pages/Admin/Charity/index.tsx:
```diff
@@ -1,34 +1,56 @@
 import Withdraws from "./Withdraws";
 import { settings } from "./routes";
 
+const LINK_GROUPS: LinkGroup[] = [
```
is it possible for type `LinkGroup` to be

```ts
export type LinkGroup = {
  title?: string;
  links: SIDEBAR_ROUTES;
};

so `<Layout/>` could just accept
```

```ts
links: [adminRoutes.index, adminRoutes.withdraws, adminRoutes.contributions]
```

misicnenad replied:
The only benefit of this approach would be that less data is copied when creating the `LINK_GROUPS` constants due to current assignment logic (e.g. `links: [LINKS.proposals]` copies the `LINKS.proposals` object into the `links` array). The drawback would be that now `Sidebar` would have `LINK_GROUPS` as a dependency, which means it would be tightly related to admin-related routes and the corresponding data. 
The benefit is not worth the drawback IMO, as the data-copying is not performance-intensive (only a handful of very small objects are copied) and is done only once, while leaving the `Sidebar` potentially re-usable.

Am I missing some other benefit of the suggested approach?

misicnenad replied:
EDIT: `Sidebar` is already dependent on the admin pages somewhat due to its `Header` component depending on `useAdminResources`, but the benefit of making an additional dependency is not that significant IMO.

ap-justin commented on src/pages/Admin/constants.ts:
```diff
@@ -40,3 +42,173 @@ export const proposalShape: SchemaShape<ProposalBase> = {
+
+const _to: keyof Link = "to";
+
+export const LINKS: { [key in SIDEBAR_ROUTES]: Link & { [_to]: key } } = {
```
nit: could convert `adminRoutes` to just `{ } as const` so could just access with "string" 
```ts
const link  = LINKS.withdraw // or ["withdraw"]
```

misicnenad replied:
Working on updating route logic to accommodate the `Liquid/Locked Account` pages.

misicnenad commented on src/constants/routes.ts:
```diff
@@ -31,18 +31,20 @@ export enum profileRoutes {
 }
 
-export enum adminRoutes {
+export const adminRoutes = {
```
By turning `adminRoutes` into a constant, it is possible to assign same values to different fields, in this case `index` and `idnex_review` have the same value `""`.
This is useful in general, but in this case it was necessary so that it becomes possible to assign different page titles to the same index path `""`.

misicnenad commented:
Updated the PR to include the most recent changes to `RC-v1.8` (done the same to the target `sidebar-admin` branch) and fixed all lingering issues (that popped up).
